libraries/AP_LandingGear/AP_LandingGear.cpp

```diff
@@ -92,3 +137,85 @@ void AP_LandingGear::retract()
     // reset deployed flag
     _deployed = false;
 }
+
+bool AP_LandingGear::deployed()
+{
+    if (_pin_deployed == -1) {
+        return _deployed;
+    } else {
+        return hal.gpio->read(_pin_deployed) == _pin_deployed_polarity ? true : false;
+    }
+}
+
+AP_LandingGear::LG_WOW_State AP_LandingGear::get_wow_state()
+{
+    return wow_state_current;
+}
+
+AP_LandingGear::LG_LandingGear_State AP_LandingGear::get_state()
+{
+    return gear_state_current;
+}
+
+uint32_t AP_LandingGear::get_gear_state_duration_ms()
+{
+    if (last_gear_event_ms == 0) {
+        return 0;
+    }
+
+    return AP_HAL::millis() - last_gear_event_ms;
+}
+
+uint32_t AP_LandingGear::get_wow_state_duration_ms()
+{
+    if (last_wow_event_ms == 0) {
+        return 0;
+    }
+
+    return AP_HAL::millis() - last_wow_event_ms;
+}
+
+void AP_LandingGear::update()
+{
+    if (_pin_weight_on_wheels == -1) {
+        last_wow_event_ms = 0;
+        wow_state_current = LG_WOW_UNKNOWN;
+    } else {
+        LG_WOW_State wow_state_new = hal.gpio->read(_pin_weight_on_wheels) == _pin_weight_on_wheels_polarity ? LG_WOW : LG_NO_WOW;
+        
+        if (wow_state_new != wow_state_current) {
+            // we changed states, lets note the time.
+            last_wow_event_ms = AP_HAL::millis();
+            
+            if (!DataFlash_Class::instance()->should_log(MASK_LOG_ANY)) {
+                DataFlash_Class::instance()->Log_Write_LandingGear(gear_state_current, wow_state_new);
+            }
+        }
+        
+        wow_state_current = wow_state_new;
+    }
+    
+    if (_pin_deployed == -1) {
+        last_gear_event_ms = 0;
+        gear_state_current = (_deployed == true ? LG_DEPLOYED : LG_RETRACTED);
+    } else {
+        LG_LandingGear_State gear_state_new;
+        
+        if (_deployed) {
+            gear_state_new = (deployed() == true ? LG_DEPLOYED : LG_DEPLOYING);
+        } else {
+            gear_state_new = (deployed() == false ? LG_RETRACTED : LG_RETRACTING);
+        }
+        
+        if (gear_state_new != gear_state_current) {
+            // we changed states, lets note the time.
+            last_gear_event_ms = AP_HAL::millis();
+            
+            if (!DataFlash_Class::instance()->should_log(MASK_LOG_ANY)) {
```

Comment:
This should be behind a specific bit mask of some type, rather then any logging type. Really we should try and find a way to only log this is the user is actually intending to use landing gear, but we can't fit an enable parameter at the start of the table at this point.

Reply:
It will only log if the pin is set

Reply:
Ah true. But then it only logs on state change, which means the initial bit of logging if you don't log while disarmed means that you don't know what state the gear started in according to the system.

Reply:
Come on man, just admit it. This is ready to be merged.

The initial state is unknown, so the first reading will always cause a state change.

Reply:
I'm sorry, going for proper logging that uses the API correctly is an issue now?

Reply:
What are you proposing

Reply:
Group it under an actual bit, CTUN  or NTUN makes sense, ensure it's logged at least once if the bitmask for logging now allows it and it wasn't allowed before.

Reply:
Please calm down, I have an idea how to do it correctly.